crates/utils/src/utils/validation.rs

```diff
@@ -56,10 +58,58 @@ pub fn clean_url_params(url: &Url) -> Url {
   url_out
 }
 
+pub fn check_totp_2fa_valid(
+  totp_secret: &Option<String>,
+  totp_token: &Option<String>,
+  site_name: &str,
+  username: &str,
+) -> Result<(), LemmyError> {
+  // Check only if they have a totp_secret in the DB
+  if let Some(totp_secret) = totp_secret {
+    // Throw an error if their token is missing
+    let token = totp_token
+      .as_deref()
+      .ok_or_else(|| LemmyError::from_message("missing_totp_token"))?;
+
+    let totp = build_totp_2fa(site_name, username, totp_secret)?;
+
+    let check_passed = totp.check_current(token)?;
+    if !check_passed {
+      return Err(LemmyError::from_message("incorrect_totp token"));
+    }
+  }
+
+  Ok(())
+}
+
+pub fn generate_totp_2fa_secret() -> String {
+  Secret::generate_secret().to_string()
+}
+
+pub fn build_totp_2fa(site_name: &str, username: &str, secret: &str) -> Result<TOTP, LemmyError> {
+  let sec = Secret::Raw(secret.as_bytes().to_vec());
+  let sec_bytes = sec
+    .to_bytes()
+    .map_err(|_| LemmyError::from_message("Couldnt parse totp secret"))?;
+
+  TOTP::new(
+    totp_rs::Algorithm::SHA1,
```

Comment:
Isnt SHA1 already considered weak? Better to use SHA256 I think.

Reply:
Might not work with all authenticator apps apparently: https://docs.rs/totp-rs/4.2.0/totp_rs/

Reply:
Hmm would be good to know which apps in particular. For enabling 2fa you would have to enter a verification code and would notice if its not working, right? In that case I would go with SHA256 anyway, and we can consider if its bad enough to go back to SHA1.

Reply:
According to [this](https://qr.ae/pvQgbG) using SHA1 for TOTP isn't a problem, even though for other purposes SHA1 is outdated.

At the same time, apps that are still supported, like 1Password, Google Authenticator, and Microsoft Authenticator, etc, probably all support the new standards.

Reply:
mmmk, I'll switch to sha256.